xcm-support/src/lib.rs

```diff
@@ -75,3 +78,75 @@ impl UnknownAsset for () {
 		Err(DispatchError::Other(NO_UNKNOWN_ASSET_IMPL))
 	}
 }
+
+/// Extracts the `AccountId32` from the passed `location` if the network
+/// matches.
+pub struct RelaychainAccountId32Aliases<Network, AccountId>(PhantomData<(Network, AccountId)>);
+impl<Network: Get<NetworkId>, AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone>
+	xcm_executor::traits::Convert<MultiLocation, AccountId> for RelaychainAccountId32Aliases<Network, AccountId>
+{
+	fn convert(location: MultiLocation) -> Result<AccountId, MultiLocation> {
+		let id = match location {
+			MultiLocation {
+				parents: 1,
+				interior: X1(AccountId32 {
+					id,
+					network: NetworkId::Any,
+				}),
+			} => id,
+			_ => return Err(location),
+		};
+		Ok(id.into())
+	}
+
+	fn reverse(who: AccountId) -> Result<MultiLocation, AccountId> {
+		Ok(AccountId32 {
+			id: who.into(),
+			network: Network::get(),
+		}
+		.into())
+	}
+}
+
+pub struct AllowDescendOrigin<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowDescendOrigin<T> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		_weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		ensure!(T::contains(origin), ());
+		let mut iter = message.0.iter_mut();
+		let i = iter.next().ok_or(())?;
+		match i {
+			DescendOrigin(..) => (),
```

Comment:
I'm struggling to see what the purpose of this is. Is it trying to allow for lowering privileges? If so, then this should check whether or not the `InteriorMultiLocation` inside of the `DescendOrigin` is `Here` or not, because a `Here` would not lower any privileges.

Would also help if there are docs detailing what `AllowDescendOrigin` seeks to achieve.

Reply:
in one of our user case(the testcase is [here](https://github.com/open-web3-stack/open-runtime-module-library/blob/relay_account/xtokens/src/tests.rs#L460)), the xcm format is: `DescendOrigin | WithdrawAsset | BuyExecution | Transact`, if the Barrier is `AllowTopLevelPaidExecutionFrom`, then it would't passing Barrier check. so we need `AllowDescendOrigin`.

I haven't consider about the interior location inside [`AllowDescendOrigin`](https://github.com/paritytech/polkadot/blob/release-v0.9.12/xcm/xcm-executor/src/lib.rs#L335-L340), in my testcase, the `who` of  `AllowDescendOrigin(who)` is `Alice` account, and the `self.origin` is `Parent`, the result `origin` is `(Parent, Alice)`, that's why we add `RelaychainAccountId32Aliases` also.

Reply:
But why is there a `DescendOrigin`? What is it trying to achieve here? To check that the origin of the XCM ultimately came from an account that the recipient controls? If so, then the name of this struct needs to reflect that, i.e. you're not simply allowing descend origin, you're allowing accounts owned by the same user to send XCMs to themselves. That sounds like `AllowEquivalentAccountsFrom`.

Moreover, there will also need to be a logic to check whether the recipient account matches the `AccountId` of the `origin` as well.

Reply:
sorry not detail the background. the xcm message cames like that: a RelayChain acount send `Transact` to a dest Parachain. As the `interior` is not `Here`, but an `Account`, so it'll add `DescendOrigin` in the sender side. that's why dest Parachain got `DescendOrigin` at first. 

> What is it trying to achieve here? To check that the origin of the XCM ultimately came from an account that the recipient controls?

yeap, so much like. in my background written above, when Parachain got the resulting origin as `(Parent, Account)`, we want the origin become to `Account` again when execute `Transact`, that's why `RelaychainAccountId32Aliases` is used, it'll make `(Parent, Account)` transform to `Account`, after that we'll use this parachain Account do our own logic. and here the final resulting `Account` is what you called _an account that the recipient controls_

to sum up, the `AllowDescendOrigin` is for barrier check, the `RelaychainAccountId32Aliases` is used for transform a parent account(sender) to account that parachain(receiver) can control.

the name `AllowDescendOrigin` is unthinkingly, you sugesst of `AllowEquivalentAccountsFrom` sounds good.
we'll rename it.

ps: I'm also thinking/consider about, we may just allow `DescendOrigin | BuyExecution | Transact` instead of `DescendOrigin | WithdrawAsset | BuyExecution | ...`, as our case may need `WithdrawAsset` after `DescendOrigin` and before `Transact`. what do you think? @xlc

Reply:
It is best to whitelist as minimal as possible. So it is best if you can reduce allowed actions as much as possible while keeping the whole e2e flow working

Reply:
I second Bryan's opinion. Casting too wide a net easily leads to a privilege escalation attack.

Reply:
unfortunately, `DescendOrigin | BuyExecution(limited weight) | Transact` can't work, unless `BuyExecution` use `Unlimited`, then It's just like `DescendOrigin | Transact`. because `BuyExecution(limited)` needs `holding` has assets, orelse it'll cause `NotHoldingFees`. I'll try `DescendOrigin | WithdrawAsset | BuyExecution(limited weight) | Transact`, and only use assets in `WithdrawAsset` for payment fees.

Reply:
@KiChjang do you think we need to add `DescendOrigin(..)` restricted condition here such as `DescendOrigin(X1(Junction::AccountId32(..))`?